pool.go

```diff
@@ -269,20 +281,42 @@ func (p *Pool[T]) Stat() *Stat {
 		}
 	}
 
-	p.cond.L.Unlock()
 	return s
 }
 
-// valueCancelCtx combines two contexts into one. One context is used for values and the other is used for cancellation.
-type valueCancelCtx struct {
-	valueCtx  context.Context
-	cancelCtx context.Context
+// tryAcquireIdleResource checks if there is any idle resource. If there is
+// some, this method removes it from idle list and returns it. If the idle pool
+// is empty, this method returns nil and doesn't modify the idleResources slice.
+//
+// WARNING: Caller of this method must hold the pool mutex!
+func (p *Pool[T]) tryAcquireIdleResource() *Resource[T] {
+	res, ok := p.idleResources.Pop()
+	if !ok {
+		return nil
+	}
+
+	res.status = resourceStatusAcquired
+	return res
 }
 
-func (ctx *valueCancelCtx) Deadline() (time.Time, bool) { return ctx.cancelCtx.Deadline() }
-func (ctx *valueCancelCtx) Done() <-chan struct{}       { return ctx.cancelCtx.Done() }
-func (ctx *valueCancelCtx) Err() error                  { return ctx.cancelCtx.Err() }
-func (ctx *valueCancelCtx) Value(key any) any           { return ctx.valueCtx.Value(key) }
+// createNewResource creates a new resource and inserts it into list of pool
+// resources.
+//
+// WARNING: Caller of this method must hold the pool mutex!
```

Comment:
It could have false negatives, but it would have no false positives, which is the important thing for such a check.

Reply:
> It could have false negatives, but it would have no false positives, which is the important thing for such a check.
True.

To be honest I don't see such a value in this check. I agree that technically it's possible and it could find some bug. But first, we all are trivially able to verify that both those calls of this function are bug free so no check that would slow real production code is necessary. For this keep in mind that atomic operation is still more costly than standard check for the panic referred above. Second reason not to do this check is that in some cases you do recursion in parallel code or multiple sub-calls and the cumulated cost of all those checks would be just too high. But this is an argument non-applicable in this situation, just a practice I try to follow.